design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+implementations provides a specific check - e.g. check if a database is up.
+
+Keycloak will provide its own implementation of a class similar to `SmallRyeHealthReporter` - `KeycloakHealthReporter`.
+Since `SmallRyeHealthReporter` is relies heavily on CDI (it uses `Instance<Healthcheck>` for injection), we need to
+provide a class with similar capabilities that enables us to assemble a centralized Health reporting instance manually.
+
+Individual Healthcheck implementations, such as database check, External IdP check or LDAP check need to be provided
+as `HealthCheck` class implementations. Later on, they will be added to `KeycloakHealthReporter`.
+
+#### Keycloak REST integration
+
+Integrating with Wildfly's `/health` endpoint requires either adding CDI annotations to Keycloak and relying on CDI
+deployment processor (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java))
+or integrating with Microprofile Health Subsystem. An additional problem is that Wildfly provides its own
+`HealthReporter` class (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporter.java))
+that doesn't use groups. The Grouping functionality is essential for this design (explained below).
+
+The biggest downside of providing Keycloak's own implementation is that we'll end up with 2 health endpoints -
+one for Wildfly (`http://localhost:9990/health`) and one for Keycloak (`https://localhost:8443/auth/health`). We
+need to be very clear in the documentation that the latter is the proper one.
+
+Having the above in mind, the design assumes integrating the new Healthcheck API with Keycloak's REST interface.
+The implementation will create additional `/health` endpoint that will parse incoming query parameters. Once all
+parameters are parsed, the implementation will extract all additional checks specified in `fail-on` parameter.
+Those checks will be reflected as groups in `KeycloakHealthReporter`. The endpoint implementation should be similar
+to Smallrye's `SmallRyeHealthGroupServlet` (see [the code](https://github.com/smallrye/smallrye-health/blob/811c63198e5876704cebf3691fe3e4dba499fa8b/testsuite/tck/src/main/java/io/smallrye/health/SmallRyeHealthGroupServlet.java#L13)).
+
+#### Healthcheck API for Liveness and Readiness probes
+
+The new Healthcheck API gives a lot of flexibility for defining both Liveness and Readiness Checks. The proposed usage
+looks like the following:
+
+    # Liveness Check
+    curl https://<keycloak-url>:8443/auth/health
+
+    # The most basic Readiness Check
+    curl https://<keycloak-url>:8443/auth/health?fail-on=db
+
+    # The more advanced Readiness Check
+    curl https://<keycloak-url>:8443/auth/health?fail-on=db,idp-google,user-federation-myldap
```

Comment:
Is there a reason it can't be defined as it's own health check group?

Reply:
If I'm reading you correctly, you're suggesting aligning with the [Groups proposal for Smallrye Health](https://xstefank.io/2020/04/21/health-groups-experimental-microprofile-health-extension-in-quarkus/). Being more specific, the call would look like this:

    curl https://<keycloak-url>:8443/auth/health/group/db
    curl https://<keycloak-url>:8443/auth/health/group/user-federation-myldap
    curl https://<keycloak-url>:8443/auth/health/group/idp-google

If you'd like to check them all, you'd call:

    curl https://<keycloak-url>:8443/auth/health/group

This is a very interesting idea. Martin told me that the Groups concept should be pushed into Microprofile Health pretty soon. That would mean we do something more or less standardized over the industry. I like this idea.

@stianst @vmuzikar @abstractj @ASzc @hmlnarik @miquelsi Perhaps you'd be interested in sharing your thoughts here?

Reply:
Not sure to be honest. First and foremost we are creating a health/readiness endpoint for K8s right. So how would you then use the groups?

Reply:
Groups use different approach to the logic than we use in our design. With our approach it is very easy to include or exclude a certain check (like a database check for example), whereas it's pretty hard to reflect this with Groups.

Let's consider the following checks for example:

    # Readiness Check - checks database and Google External IdP
    https://<keycloak-url>:8443/auth/health?fail-on=db,idp-google
    # Liveness Check - checks only if REST interface is up
    https://<keycloak-url>:8443/auth/health

We could certainly implement this with separate `https://<keycloak-url>:8443/auth/health/ready` and `https://<keycloak-url>:8443/auth/health/live` but let's try to focus on Groups here. The `/health/group` REST interface aggregates all groups. This means it would check the database, External IdPs, LDAP... everything. In our example, we'd like to exclude all External IdP apart from Google as well as all User Federation checks (LDAP etc). The current SmallRye Groups implementation seems to allow calling a single Group per invocation. That means, we'd need to call the probe two times:

     https://<keycloak-url>:8443/auth/health/groups/db
     https://<keycloak-url>:8443/auth/health/groups/idp-google

This is a bit unfortunate, as Kubernetes allows only one HTTP Probe definition (although, we could craft a Bash script up to call both endpoints for us). 

Calling the Liveness example is also problematic, because we want to exclude all checks there. We just want to ping the REST interface and see if its up. The only idea I have in mind is to create a separate Group for it - something like:

    https://<keycloak-url>:8443/auth/health/groups/rest-only

But again, all this seems to be less flexible than the proposal from this Design. 

Let me also hook @xstefank into the discussion. Perhaps he has some neat idea that we could use here.

Reply:
@slaskawi you can only define another group that will combine both `db` and `idp-google` groups into a single URL as mentioned. Health checks can be part of several groups. However, it might be useful to have also exclude options if you have many groups/checks. This is one of the most complex use cases I've seen so far so I would like to understand what is your preferred approach so we can adjust the feature. If you feel like that `fail-on` or `exclude` would be better for you please create an issue in the smallrye-health repo describing the use case.

Reply:
Thanks you for the input @xstefank. Together with @stianst we decided to simplify the design and leave the idea of the optional checks out.

Reply:
Let me resolve this conversation.